src/main/java/io/cryostat/messaging/notifications/NotificationListener.java

```diff
@@ -0,0 +1,47 @@
+/*
+ * Copyright The Cryostat Authors
+ *
+ * The Universal Permissive License (UPL), Version 1.0
+ *
+ * Subject to the condition set forth below, permission is hereby granted to any
+ * person obtaining a copy of this software, associated documentation and/or data
+ * (collectively the "Software"), free of charge and under any and all copyright
+ * rights in the Software, and any and all patent rights owned or freely
+ * licensable by each licensor hereunder covering either (i) the unmodified
+ * Software as contributed to or provided by such licensor, or (ii) the Larger
+ * Works (as defined below), to deal in both
+ *
+ * (a) the Software, and
+ * (b) any piece of software and/or hardware listed in the lrgrwrks.txt file if
+ * one is included with the Software (each a "Larger Work" to which the Software
+ * is contributed by such licensors),
+ *
+ * without restriction, including without limitation the rights to copy, create
+ * derivative works of, display, perform, and distribute the Software and make,
+ * use, sell, offer for sale, import, export, have made, and have sold the
+ * Software and the Larger Work(s), and to sublicense the foregoing rights on
+ * either these or other terms.
+ *
+ * This license is subject to the following condition:
+ * The above copyright notice and either this complete permission notice or at
+ * a minimum a reference to the UPL must be included in all copies or
+ * substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+package io.cryostat.messaging.notifications;
+
+import java.util.Set;
+
+import io.vertx.core.eventbus.Message;
+
+public interface NotificationListener {
```

Comment:
What if this was like:

```java
public interface NotificationListener<T> {
  void onMessage(String category, Message<T> message);
}
```

Reply:
This has the problem that it will enforce that all of the messages have the same type across all of the categories in the `Set`, though.

Reply:
Yeah, and then I get problems trying to cast the `Message<Object>` message callback object into the `Message<T>` `onMessage` arg. I think maybe this problem could be solved if we used [MessageCodec](https://vertx.io/docs/apidocs/io/vertx/core/eventbus/MessageCodec.html)? I think then we could just use it to register codecs for each listener `onMessage` `T` type and it won't complain.

Reply:
Sure, that looks like the right solution if we need to send generalized data types as the message.

> Usually the event bus only allows a certain set of message types to be sent across the event bus, including primitive types, boxed primitive types, byte[], [JsonObject](https://vertx.io/docs/apidocs/io/vertx/core/json/JsonObject.html), [JsonArray](https://vertx.io/docs/apidocs/io/vertx/core/json/JsonArray.html), [Buffer](https://vertx.io/docs/apidocs/io/vertx/core/buffer/Buffer.html).

That sounds like it might already cover our needs without having to implement a MessageCodec, though. Our notification messages generally just contain a simple serialized `Map`, which is more or less the same thing as a `JsonObject`, so perhaps we should just orient the interface toward that instead. ie rather than being generic, require the message type to be `JsonObject`.

Reply:
Also, I would prefer to not expose the `Message` class from Vertx through the `NotificationListener` interface. Part of the design goal of this interface is to abstract the Vertx EventBus away, so listeners and publishers shouldn't know anything about Vertx, the EventBus, or the Message type.

The EventBus and Message have a lot more capability after all, and in this use case we're just trying to use one narrow facet.

```java
public interface NotificationListener<T> {
  void onMessage(String category, T message);
}
```

Reply:
In the publisher, it would look something like:

```java
        listeners.forEach(listener -> {
            listener.categories().forEach(category -> {
                vertx.eventBus().consumer(category, message -> listener.onMessage(category, message.body()));
            });
        });
```

Reply:
I'm not sure how to get past these compile time errors 

If I do this 
```
public interface NotificationListener<T> {
  void onMessage(String category, T message);
}
...
...
...

    private final Vertx vertx;
    private final Set<NotificationListener<?>> listeners;

listeners.forEach(listener -> {
            listener.categories().forEach(category -> {
                vertx.eventBus().consumer(category, message -> listener.onMessage(category, message.body()));
            });
        });
```
It won't like it since `message.body()` is of type Object and onMessage expects type `<T>`, but since the `Set` is a set of `NotificationListener<?>`, it won't accept any casting at runtime. 

So instead should `onMessage` be like 
`void onMessage(String category, JsonObject msg);`
And we parse it in the receiver?

```
    NotificationPublisher(Vertx vertx, Set<NotificationListener<?>> listeners) {
        this.vertx = vertx;
        this.listeners = listeners;

        listeners.forEach(listener -> {
            listener.categories().forEach(category -> {
                vertx.eventBus().consumer(category, message -> {
                    JsonObject msg = JsonObject.mapFrom(message.body());
                    listener.onMessage(category, msg);
                });
            });
        });
    }
```

Reply:
That is an option but it's really unfortunate if we have to serialize and deserialize JSON for this. I would probably rather say forget the Vertx EventBus in that case and just implement a plain listener/callback pattern in plain Java, in memory.

It's a common pattern, and we have an example already here:

https://github.com/cryostatio/cryostat/blob/5fae352f0fd66454bab564844be5943c83fba7c0/src/main/java/io/cryostat/platform/PlatformClient.java#L51

https://github.com/cryostatio/cryostat/blob/5fae352f0fd66454bab564844be5943c83fba7c0/src/main/java/io/cryostat/platform/internal/AbstractPlatformClient.java#L67

This is for the `TargetDiscoveryEvent`, but it should generalize to `Notification<T>` fairly easily. The `Set<String> categories()` part would still be a problem, so maybe we just drop that concept for now.

Reply:
Thanks. Sorry for wasting your effort on the EventBus - hopefully you at least learned some new patterns or techniques through that experience.